### PR TITLE
[TRUS-3183] - Enrolling a Lead Trustee after registering a non-taxable trust

### PR DESCRIPTION
**Pull request by jassalrichy**
Implement a custom callback endpoint for non-taxable trusts
Rename the current callback endpoint for taxable trusts (4MLD)
Going to switch the service name and callback endpoint provided to tax-enrolments if we know they're taxable or not

### CODE REVIEW

File: app/config/AppConfig.scala [hunk 1]
```diff
@@ -37,14 +37,21 @@ class AppConfig @Inject()(configuration: Configuration, servicesConfig: Services
   val variationsApiSchema4MLD: String = "/resources/schemas/4MLD/variations-api-schema-4.0.json"
   val variationsApiSchema5MLD: String = "/resources/schemas/5MLD/variations-api-schema-4.2.0.json"
 
+  private def insertTRN(url: String, trn: String) = url.replace(":trn", trn)
+
   val taxEnrolmentsPayloadBodyServiceName : String =
     configuration.get[String]("microservice.services.tax-enrolments.serviceName")
```
Comment (adamconder) [inline app/config/AppConfig.scala:43]: Could you make this path similar to non-taxable and include taxable in the path?

Comment (adamconder) [reply app/config/AppConfig.scala:43]: I'd also suggest adding taxable into the value name

File: app/connector/TaxEnrolmentConnector.scala [hunk 2]
```diff
@@ -39,28 +39,39 @@ class TaxEnrolmentConnectorImpl @Inject()(http: HttpClient,
...
-  override  def enrolSubscriber(subscriptionId: String)(implicit hc: HeaderCarrier) :  Future[TaxEnrolmentSuscriberResponse] = {
+  override  def enrolSubscriber(subscriptionId: String, taxable: Boolean, trn: String)(implicit hc: HeaderCarrier) :  Future[TaxEnrolmentSuscriberResponse] = {
     for {
-      serviceName <- getServiceName
-      response <- getResponse(subscriptionId, serviceName)
+      is5MLD <- is5MLD
+      response <- getResponse(subscriptionId, is5MLD, taxable, trn)
     } yield
     response
   }
```
Comment (adamconder) [inline app/connector/TaxEnrolmentConnector.scala:77]: We may want to handle the call to get is5MLD failing and recovering with an audit event.

File: app/controllers/RegisterTrustController.scala [hunk 3]
```diff
@@ -82,7 +82,10 @@ class RegisterTrustController @Inject()(desService: DesService, config: AppConfi
                         response = response
                       )
 
-                      rosmPatternService.enrolAndLogResult(response.trn, request.affinityGroup) map {
+                      rosmPatternService.enrolAndLogResult(response.trn,
+                        request.affinityGroup,
+                        trustsRegistrationRequest.trust.details.trustTaxable.getOrElse(false)
```
Comment (adamconder) [inline app/controllers/RegisterTrustController.scala:87]: Wouldn't this be .getOrElse(true)?

It won't be present for current trusts being registered in 4MLD therefore we need to set taxable to true?

File: app/controllers/TaxEnrolmentCallbackController.scala [hunk 4]
```diff
@@ -32,7 +32,7 @@ class TaxEnrolmentCallbackController @Inject()(
                                                 cc: ControllerComponents
                                                ) extends BackendController(cc) with Logging {
 
-  def subscriptionCallback() = Action.async(parse.json) {
+  def subscriptionCallback(trn: String) = Action.async(parse.json) {
```
Comment (adamconder) [inline app/controllers/TaxEnrolmentCallbackController.scala:35]: Could we rename the method to me taxableSubscriptionCallback?

Comment (adamconder) [reply app/controllers/TaxEnrolmentCallbackController.scala:35]: Maybe also include that it was taxable in the log as well?

File: conf/application.conf [hunk 5]
```diff
@@ -204,9 +204,12 @@ microservice {
         tax-enrolments {
             host=localhost
             port=9802
-            serviceName5MLD="serviceName"
+            non-taxable {
+                serviceName="serviceName"
+                callback="http://localhost:9782/trusts/tax-enrolment/registration/non-taxable/hmrc-tersnt-org/:trn/subscriptions"
+            }
             serviceName="serviceName"
```
Comment (adamconder) [inline conf/application.conf:211]: Could this move into an object of it's own to match non-taxable?

File: test/services/TaxEnrolmentsServiceSpec.scala [hunk 6]
```diff
@@ -38,40 +38,43 @@ class TaxEnrolmentsServiceSpec extends BaseSpec {
...
-  ".setSubscriptionId" should {
+  ".setSubscriptionId 4MLD" should {
+
+    val taxable: Boolean = false
+    val trn = "XTRN1234567"
 
     "return TaxEnrolmentSuccess  " when {
       "connector returns success taxEnrolmentSuscriberResponse." in {
-        when(mockConnector.enrolSubscriber("123456789")).
+        when(mockConnector.enrolSubscriber("123456789", taxable, trn)).
```
Comment (adamconder) [inline test/services/TaxEnrolmentsServiceSpec.scala:48]: Are we missing specs for 5MLD?

File: test/services/RosmPatternServiceSpec.scala [hunk 7]
```diff
@@ -32,16 +32,19 @@ class RosmPatternServiceSpec extends BaseSpec {
 
   private val SUT = new RosmPatternServiceImpl(mockDesService, mockTaxEnrolmentsService)
 
-  ".completeRosmTransaction" should {
+  ".completeRosmTransaction 4MLD" should {
+
+    val taxable: Boolean = false
+    val trn = "XTRN1234567"
```
Comment (adamconder) [inline test/services/RosmPatternServiceSpec.scala:38]: Are we missing specs for 5MLD?

File: test/connectors/TaxEnrolmentConnectorSpec.scala [hunk 8]
```diff
@@ -40,7 +40,10 @@ class TaxEnrolmentConnectorSpec extends ConnectorSpecHelper {
       )
   }
 
-  ".enrolSubscriber" should {
+  ".enrolSubscriber 4MLD" should {
```
Comment (adamconder) [inline test/connectors/TaxEnrolmentConnectorSpec.scala:43]: Missing specs for 5MLD?

**Review by adamconder — CHANGES_REQUESTED** [review]

File: app/config/AppConfig.scala (hunk 1 above)
Comment (jassalrichy) [reply app/config/AppConfig.scala:43]: Done

File: app/connector/TaxEnrolmentConnector.scala (hunk 2 above)
Comment (jassalrichy) [reply app/connector/TaxEnrolmentConnector.scala:77]: Done

File: app/controllers/RegisterTrustController.scala (hunk 3 above)
Comment (jassalrichy) [reply app/controllers/RegisterTrustController.scala:87]: Done

File: app/controllers/TaxEnrolmentCallbackController.scala (hunk 4 above)
Comment (jassalrichy) [reply app/controllers/TaxEnrolmentCallbackController.scala:35]: Done

File: conf/application.conf (hunk 5 above)
Comment (jassalrichy) [reply conf/application.conf:211]: Done

File: test/services/TaxEnrolmentsServiceSpec.scala (hunk 6 above)
Comment (jassalrichy) [reply test/services/TaxEnrolmentsServiceSpec.scala:48]: Updated but getting failed tests when mocking config returns

File: test/services/RosmPatternServiceSpec.scala (hunk 7 above)
Comment (jassalrichy) [reply test/services/RosmPatternServiceSpec.scala:38]: tests updated to pass taxable flag and trn

File: test/connectors/TaxEnrolmentConnectorSpec.scala (hunk 8 above)
Comment (jassalrichy) [reply test/connectors/TaxEnrolmentConnectorSpec.scala:43]: Updated but getting failed tests when mocking config returns

File: test/services/TaxEnrolmentsServiceSpec.scala (hunk 6 above)
Comment (jassalrichy) [reply test/services/TaxEnrolmentsServiceSpec.scala:48]: tests updated to pass taxable flag and trn

**Review by adamconder — APPROVED** [review]

**Pull request merged**